main.go

```diff
@@ -0,0 +1,171 @@
+			log.Printf("Redirecting /api/v0 to %s", strings.Join(kuboRPC, " "))
+			err := gatewaySrv.ListenAndServe()
+			if err != nil && !errors.Is(err, http.ErrServerClosed) {
+				log.Printf("Failed to start gateway: %s", err)
+				quit <- os.Interrupt
+			}
+		}()
+
+		go func() {
+			defer wg.Done()
+			log.Printf("Metrics listening on http://127.0.0.1:%d", metricsPort)
+			err := metricsSrv.ListenAndServe()
+			if err != nil && !errors.Is(err, http.ErrServerClosed) {
+				log.Printf("Failed to start metrics: %s", err)
+				quit <- os.Interrupt
+			}
+		}()
+
+		signal.Notify(quit, os.Interrupt)
+		<-quit
+		log.Printf("Closing servers...")
+		go gatewaySrv.Close()
+		go metricsSrv.Close()
+		wg.Wait()
+		return nil
+	},
+}
+
+func makeGatewayHandler(saturnOrchestrator, saturnLogger string, kuboRPC []string, port int) (*http.Server, error) {
+	blockStore, err := newBlockStore(saturnOrchestrator, saturnLogger)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sets up an offline (no exchange) blockService based on the Saturn block store.
+	blockService := blockservice.New(blockStore, offline.Exchange(blockStore))
+
+	// // Sets up the routing system, which will proxy the IPNS routing requests to the given gateway.
+	routing := newProxyRouting(kuboRPC, nil)
+
+	// Creates the gateway with the block service and the routing.
+	gwAPI, err := newBifrostGateway(blockService, routing)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := map[string][]string{}
+	gateway.AddAccessControlHeaders(headers)
+
+	gwConf := gateway.Config{
+		Headers: headers,
+	}
+
+	gwHandler := gateway.NewHandler(gwConf, gwAPI)
+	mux := http.NewServeMux()
+	mux.Handle("/ipfs/", gwHandler)
+	mux.Handle("/ipns/", gwHandler)
+	mux.Handle("/api/v0/", newAPIHandler(kuboRPC))
+
+	// Note: in the future we may want to make this more configurable.
+	noDNSLink := false
+	publicGateways := map[string]*gateway.Specification{
+		"localhost": {
+			Paths:         []string{"/ipfs", "/ipns"},
+			NoDNSLink:     noDNSLink,
+			UseSubdomains: true,
+		},
+		"dweb.link": {
+			Paths:         []string{"/ipfs", "/ipns"},
+			NoDNSLink:     noDNSLink,
+			UseSubdomains: true,
+		},
+	}
+
+	return &http.Server{
+		Handler: gateway.WithHostname(mux, gwAPI, publicGateways, noDNSLink),
+		Addr:    ":" + strconv.Itoa(port),
+	}, nil
+}
+
+func makeMetricsHandler(port int) (*http.Server, error) {
+	mux := http.NewServeMux()
+	mux.Handle("/debug/metrics/prometheus", promhttp.HandlerFor(prometheus.DefaultGatherer, promhttp.HandlerOpts{}))
+
+	return &http.Server{
+		Handler: mux,
+		Addr:    ":" + strconv.Itoa(port),
+	}, nil
+}
+
+func newAPIHandler(endpoints []string) http.Handler {
+	s := rand.NewSource(time.Now().Unix())
+	rand := rand.New(s)
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Naively choose one of the Kubo RPC clients.
+		endpoint := endpoints[rand.Intn(len(endpoints))]
+		http.Redirect(w, r, endpoint+r.URL.Path+"?"+r.URL.RawQuery, http.StatusFound)
```

Comment:
:information_source:  this will be refined in https://github.com/ipfs/bifrost-gateway/issues/13